File: DAYS/Day53-Minimum window substring/minWindow.py
# Strings : Minimum Window Substring (LeetCode 76)
# Given two strings s and t, return the minimum window substring of s such that every character
# in t (including duplicates) is included in the window. If there is no such substring, return "".
# Example 1: Input: s = "ADOBECODEBANC", t = "ABC" Output: "BANC"
# Example 2: Input: s = "a", t = "a" Output: "a"
# Example 3: Input: s = "a", t = "aa" Output: ""
# Different Approaches:
# 1. Brute Force - time O(n^3) space O(1)
# 2. Sliding Window - time O(n) space O(1)
# Solution 2: Sliding Window
# Leetcode Link: https://leetcode.com/problems/minimum-window-substring/description/

class Solution:
    def minWindow(s,t):
        if t=="":return ""
    @staticmethod
    def minWindow(s: str, t: str) -> str:
        if not t:
            return ""
        # frequency map for characters in t
        count = {}
        for ch in t:
            count[ch] = count.get(ch, 0) + 1
        window = {}
        l = 0
        have = 0
        need = len(count)
        res = [-1, -1]
        resLen = float("inf")
        # Expand window with right pointer r
        for r, ch in enumerate(s):
            window[ch] = window.get(ch, 0) + 1
            if ch in count and window[ch] == count[ch]:
                have += 1
            # Try to contract from the left while we still have a valid window
            while have == need:
                if (r - l + 1) < resLen:
                    resLen = r - l + 1
                    res = [l, r]
                # remove leftmost char from window
                window[s[l]] -= 1
                if s[l] in count and window[s[l]] < count[s[l]]:
                    have -= 1
                l += 1
        if resLen == float("inf"):
            return ""
        return s[res[0]: res[1] + 1]
# Test Cases
print(Solution.minWindow("ADOBECODEBANC","ABC")) # Output: "BANC"
print(Solution.minWindow("a","a")) # Output: "a"    
print(Solution.minWindow("aaabbaaabbbcc","abc")) # Output: "abbbc"

# Best Approach : Sliding Window
# Complexity Analysis: Time Complexity: O(n) , Space Complexity: O(1)

File: DAYS/Day53-Minimum window substring/minWindow.js
// Strings : Minimum Window Substring leetcode 76
// Given two strings s and t of lengths m and n respectively, return the minimum window substring of s such that every character in t (including duplicates) is included in the window. If there is no such substring, return the empty string "".
// The testcases will be generated such that the answer is unique.
// Example 1: Input: s = "ADOBECODEBANC", t = "ABC" Output: "BANC"
// Example 2: Input: s = "a", t = "a" Output: "a"
// Example 3: Input: s = "a", t = "aa" Output: "" Explanation: Both 'a's from t must be included in the window. Since the largest window of s only has one 'a', return empty string.

// Different Approaches
// 1. Brute Force - Time O(n^3) Space O(1)
// 2. Sliding Window - Time O(n) Space O(1)
// 3. Optimized Sliding Window - Time O(n) Space O(1)

// Solution - Sliding Window
function minWindow(s, t) {
    if(s.length<t.length) return "";
    const count=new Map() , window=new Map();
    for (let char of t){
        count.set(char,(count.get(char) || 0)+1);
    }
    let have=0,need=count.size;
    let left=0,right=0;
    let res=[-1,-1];
    let resLen=Infinity;
    for(let right=0;right<s.length;right++){
        const char=s[right];
            window.set(char,(window.get(char)||0)+1);
            if(count.has(char) && window.get(char)===count.get(char)){
                have++;
            }
            while(have===need){
                if((right-left+1)<resLen){
                    res=[left,right];
                    resLen=right-left+1;
                }
            const leftChar=s[left];
            window.set(leftChar,window.get(leftChar)-1);
            if(count.has(leftChar) && window.get(leftChar)<count.get(leftChar)){
                have--;
            }
            left++;
        }
    }
    const [l,r]=res;
    return l===-1 ? "" : s.substring(l,r+1);
}
// Test Cases
console.log(minWindow("ADOBECODEBANC", "ABC")); // Output: "BANC"
console.log(minWindow("a", "a")); // Output: "a"
console.log(minWindow("a", "aa")); // Output: ""

// Best Approach : Sliding Window
// Complexity Analysis: Time Complexity: O(n) , Space Complexity: O(1)